feat: add dynamically configured Hubble metrics

rectified95 opened the pull request:
Following the example of the Hubble Dynamic Exporter, this PR implements the CFP to allow Hubble metrics to be dynamically configured with a ConfigMap, rather than a static string at startup: https://github.com/cilium/design-cfps/blob/main/hubble/CFP-30788-managing-hubble-metrics-cardinality.md

This PR has been scoped to allow turning metric handlers on/off and applying filters - changing the label set of a registered metric is moved to a follow-up change pending more design, because of constraints in Prometheus.

### Changes:
- add dynamic flow processor: registers/unregisters metric handlers based on received config
- add metric cfg watcher: parse yaml config into MetricConfig struct added in https://github.com/cilium/cilium/pull/34849
- add FlowFilters to metrics handlers so they're only selectively applied
- update `launchHubble` to conditionally use static/dynamic metrics config
- add new ConfigMap for metric setup
- extend the `Handler` interface with `Deinit` and `HandleConfigurationUpdate` functions
- remove `FlowProcessor` interface and move `ProcessFlow` to `Handler`: all metric handlers are `FP`s and the only `FP` that's not a handler is `DropEventEmitter`, which is registered separately; `struct Handlers` is thus no longer necessary
- convert handler bulk creation functions to operate on a single handler and be iterated on


### Testing:
- yaml config read&validate: `TestReadMetricConfigFromCM`
- metric config reload, check handler set for added/removed metrics: `TestHandlersUpdatedInDfpOnConfigChange`
- pull metrics from Hubble server: `TestHubbleServerWithDynamicMetrics`
- handler re-registration + gather metrics: `TestMetricReRegisterAndCollect` 
- manually tested updating CM: hit breakpoints in new code paths under debugger
- manually collected Hubble metrics specified in CM

chancez commented on daemon/cmd/hubble.go:
```diff
@@ -191,16 +191,38 @@ func (d *Daemon) launchHubble() {
-			"metrics": option.Config.HubbleMetrics,
-			"tls":     option.Config.HubbleMetricsServerTLSEnabled,
-		}).Info("Starting Hubble Metrics server")
+		if option.Config.HubbleDynamicMetricConfigFilePath != "" {
```
We probably want to think about how we might do validation if the user tries to configure both. Potentially log an error and/or do not configure metrics? CC @kaworu

rectified95 replied:
Fow now, changed it to skip metric configuration altogether if both types are specified.

chancez commented on pkg/hubble/metrics/dynamic_flow_processor.go:
```diff
@@ -0,0 +1,134 @@
+	// mutex protects from concurrent modification of managedFlowProcessors by config
+	// reloader when hubble events are processed
+	mutex    lock.RWMutex
+	Metrics  *api.Handlers // TODO add getnames for testing encapsulation?
```
It might make sense to use a different type for dynamic metrics.  A `map[string]api.NamedHandler` where the key is the name of the handler. Though we would need to figure out how to handle `flowProcessors`. Your already creating one in the `onConfigReload` and it could simplify some of the logic to add/remove/update the handlers.

rectified95 replied:
I considered this, but figured that config change being much less frequent than calls to `ProcessFlow`, iterating over a slice is faster than a map.

As for FlowProcessor, my current read of the code is that "all metric handlers are FlowProcessors and the only FP that's not a handler is DropEventEmitter", so unless there are going to be new non-FP handlers, perhaps ProcessFlow could be moved to the Handler interface.

chancez replied:
That might make sense. For `DropEventEmitter` how would it change?

rectified95 replied:
We can still register the `DropEventEmitter` with `WithOnDecodedFlowFunc(func(ctx context.Context, flow *flowpb.Flow)` just as now - what matters is the function signature, so its `ProcessFlow` can remain unchanged (it just wouldn't be implementing an interface) or be renamed.

rectified95 replied:
Pushed two commits - one removing the `FlowProcessor` interface, and another one on top of it removing the `Handlers` interface, which becomes an unnecessary array wrapper over `[]NamedHandler`.

 Nothing changes with  `DropEventEmitter`.

chancez commented on pkg/hubble/metrics/dynamic_flow_processor.go:
```diff
@@ -0,0 +1,134 @@
+				continue
+			} else {
+				if h, ok := curHandlerMap[cm.Name]; ok {
+					h.Handler.Deinit(d.registry)
```
This might have some trade-offs we need to consider. It might make sense to change the handlers API so that each handler needs to support reconfiguration.

rectified95 replied:
Can you elaborate on this?

chancez replied:
What I mean is the API here might be wrong. Instead of `Deinit()`, it might make more sense to change to something like `HandleConfigurationUpdate(...)` instead. It feels like updates to config shouldn't be treated the same as "removed" + "re-added", because metrics are stateful data we don't want to reset ideally.

rectified95 replied:
1. Making sure you meant just the update scenario here - I think `Deinit()` still makes sense for removing a metric.

2. Whether `metrics are stateful data we don't want to reset ideally.` could depend on which scenario will be more common: 
- user applies a filter on a metric to remove noise from uninteresting pods: here we could free up resources by removing old metrics by re-registering the handler
- user only temporarily applies a filter and removes it later on, wants to see metric state from before

3. I can add `HandleConfigurationUpdate` - since the PR doesn't support label changes, this function would basically change the filter fields on the handlers, right?

chancez replied:
If you don't want to handle label updates in this PR, that's fine, but I still believe updates should be treated differently from deinit + reinit. In general, I want to bias towards not re-initializing the metrics. 

Since this PR won't support label updates, it will also need to add validation to prevent attempted updates to labels.

rectified95 replied:
Validation function updated, test case added

chancez commented on pkg/hubble/metrics/dynamic_flow_processor.go:
```diff
@@ -0,0 +1,134 @@
+	return dynamicFlowProcessor
+}
+
+func (d *DynamicFlowProcessor) onConfigReload(ctx context.Context, isSameHash bool, hash uint64, config api.Config) {
```
One thing I think this needs to consider is what we do when a bad configuration is loaded. Preferably, I think we wouldn't do anything if the new config is invalid in some way, as it might lead from a valid configuration to partially configuration loaded and the rest of the config unloaded. [I recently had to fix a few bugs in the hubble/exporter](https://github.com/cilium/cilium/commit/26331e7112188c0d24e53b6b259f3ede3d3b1630) with this. 

We may also want metrics to track metrics being reconfigured, similar to the dynamic flow exporter.

rectified95 replied:
> we wouldn't do anything if the new config is invalid in some way

Do you mean that this function should no-op if any metric config is malformed in the parsed config set, or that a single handler should be skipped during reload if the corresponding new config is invalid?

I will add some validation of the parsed struct, but if you mean the former, then a few valid configs can get registered with Prometheus, and then one will be invalid and we'd need undo the earlier ones.

rectified95 replied:
In the commit you linked, I can see a single exporter is now skipped

rectified95 replied:
Skipping whole update if validation fails.

chancez commented on pkg/hubble/metrics/metric_config_watcher.go:
```diff
@@ -0,0 +1,114 @@
+	c.logger.Debug("Attempting reload")
+	config, isSameHash, hash, err := c.readConfig()
+	if err != nil {
+		c.logger.Warnf("failed reading dynamic exporter config")
```
You're not logging the error. You may want to review my [recent PR fixing some bugs in the dynamic flow exporter](https://github.com/cilium/cilium/pull/35069/) to make sure anything your copying doesn't copy the bugs that were fixed there.

chancez commented on pkg/hubble/metrics/metric_config_watcher.go:
```diff
@@ -0,0 +1,114 @@
+
+func validateMetricConfig(config *api.Config) error {
+	return nil
+}
```
We definitely are going to want this.

rectified95 replied:
Added a check for missing metric names, duplicate metrics, and changed label set (keep track of previously registered configs in a map)

chancez commented on pkg/hubble/metrics/flow/handler.go:
```diff
@@ -8,24 +8,39 @@ import (
+	flows     *prometheus.CounterVec
+	context   *api.ContextOptions
+	cfg       *api.MetricConfig
+	AllowList filters.FilterFuncs
+	DenyList  filters.FilterFuncs
```
Double check whether or not the handler gets logged, and if so, these public struct fields aren't loggable, and will cause confusing error messages to users. See https://github.com/cilium/cilium/pull/35069/commits/9f6cddf7b9ee61eb7db0bd2ede10f1d5378fd86d for reference

rectified95 replied:
It currently only logs one field - plugin name - apart from the error it'd get from prometheus or elsewhere.

chancez commented on pkg/hubble/metrics/flow/handler.go:
```diff
@@ -92,3 +111,7 @@ func (h *flowHandler) ProcessFlow(ctx context.Context, flow *flowpb.Flow) error
 }
+
+func (h *flowHandler) Deinit(registry *prometheus.Registry) {
+	registry.Unregister(h.flows)
```
I figured I would just confirm what Prometheus does when we unregister, and it might not work exactly as expected: https://github.com/prometheus/client_golang/issues/203. We should make sure to test the case where a handler is registered, de-registered, then registered again, and verify what metrics are exported if we scrape the metrics server between each register/deregister/register step.

chancez replied:
We should also check the error from `Unregister` and if it doesn't deregister we may want to return an error on `Deinit` since it would mean re-registering in the future could potentially panic.

rectified95 replied:
Added the following passing test: re-register a deregistered handler with the same config; after dereg the corresponding metrics are gone, and after re-reg they're collected again on ProcessFlow.

Looking at `Unregister` it only returns `false` if it's called on a registry that doesn't contain a record of registration of the handler being deregistered, in which case we should still remove the handler from our list regardless of the error. Still, I changed the `Deinit` signature to return bool if implementations want to report other failures eg. due to state cleanup failure.

chancez reviewed (changes requested):
We need to be pretty careful about registration. This definitely will need some more robust tests that go through registering/deregistering/registering the same handler to verify this doesn't panic. 

You will also want to test registering/registering/registering the same handler but with different configuration too, since we want to be able to change which labels it has, but I suspect there's going to be cases where that's not valid. For example it's my understanding that it's not valid to have a metric with 3 labels, then reconfigure it to only have 2 labels, at least with the default Prometheus registry. 

Prometheus might be a bit more opinionated about this, and while it might not be valid with the default registry, we may still want to find a way to support it somehow, since it's currently possible to change labels by restarting with a different configuration. This could definitely complicate the implementation though, so I would write tests first, and see what happens, then we can figure out how to proceed. I think it's okay if the initial implementation has some limitations on reconfiguring labels, and we can come back with a custom registry implementation in a follow up PR if we need to.

We also will need more robust validation and we need to be careful about accidentally loading configurations that are invalid. I believe the correct behavior when a new config has any invalid constructs, is to not change the existing metrics configuration at all until the entire new configuration is correct.

rectified95 commented on pkg/hubble/metrics/metrics_test.go:
```diff
@@ -123,3 +101,259 @@ func TestHubbleServerStandalone(t *testing.T) {
+	assert.Nil(t, metricFamilies)
+
+	// Re-register drop handler with same config and collect metrics.
+	watcher = metricConfigWatcher{configFilePath: "testdata/valid_metric_config.yaml"}
```
This test shows a working scenario: re-registering a handler with the exact same config.

I've also tried 2 other cases: re-registration with more context options, and with more labels for the same context option set, and Prometheus panics with this error:
`"a previously registered descriptor with the same fully-qualified name as [...] has different label names or a different help string"` (vendor/github.com/prometheus/client_golang/prometheus/registry.go@315)

rectified95 replied:
As for the scenario of re-registering with different labels - could we create a new `PedanticRegistry` and re-register all handlers in it?

chancez replied:
I don't think that will work because it would mean dropping the values of existing metrics that are unchanged.

rectified95 replied:
I want to put the changes to support label updates in a follow-up PR, but use this thread to explore the solutions.
What if put a warning in the docs that changing labels restarts all handlers (equivalent to the current state of restarting cilium with a different static config)?
Do you know what a custom registry implementation would involve to support this?

chancez replied:
That feels like an unexpected behavior and a potential footman. I'd rather see if we can explore ways to support label updates without those trade-offs.  As far as solving it with a custom registry, that's probably the correct solution, but I'm not sure. A custom collector might work too, but it means changing the metrics handlers more.

chancez commented on pkg/hubble/metrics/dynamic_flow_processor.go:
```diff
@@ -0,0 +1,137 @@
+func (d *DynamicFlowProcessor) onConfigReload(ctx context.Context, isSameHash bool, hash uint64, config api.Config) {
+	if isSameHash {
+		return
+	}
```
This could probably be moved into the caller and the argument can be removed.

rectified95 replied:
done

chancez commented on pkg/hubble/metrics/http/handler.go:
```diff
@@ -196,3 +221,7 @@ func observerObserve(o prometheus.Observer, value float64, traceID string) {
 }
+
+func (h *httpHandler) Deinit(registry *prometheus.Registry) bool {
+	return registry.Unregister(h.requests) || registry.Unregister(h.responses) || registry.Unregister(h.duration)
```
An error in one unregistration means the others do not get unregistered. This doesn't seem desirable. Perhaps you should accumulate the errors instead.

rectified95 replied:
Fixed

chancez commented on pkg/hubble/metrics/api/api.go:
```diff
@@ -64,56 +58,57 @@ type Handler interface {
-type FlowProcessor interface {
+	// Deinit deregisters the metrics from the Prometheus registry
+	// and cleans up internal handler state
+	Deinit(registry *prometheus.Registry) bool
```
`Deinit` should probably return an error instead of a bool

rectified95 replied:
The underlying function doesn't return an error and only returns false when trying to unregister a non-existent collector.
`func (r *Registry) Unregister(c Collector) bool {` 
Should I change it regardless?

chancez replied:
I was thinking because of https://github.com/cilium/cilium/pull/35185#discussion_r1815581863 we should return an error indicating what failed. Eg: if we have many metrics, we should specify which failed to unregister so it can be debugged.

rectified95 replied:
Update to return error, not bool

chancez reviewed (commented):
I'm fine with not handling label changes, but in any solution of this feature, I'd like to avoid "updates" from unregistering and reseting metrics.

chancez commented on pkg/hubble/metrics/dynamic_flow_processor.go:
```diff
@@ -0,0 +1,139 @@
+		//   no-op, if config unchanged;
+		//   update handler config, if changed.
+		if m, ok := curHandlerMap[cm.Name]; ok {
+			if reflect.DeepEqual(*m.MetricConfig, *cm) {
```
I'm not sure if reflect.DeepEqual will work correctly for `*pb.FlowFilter`. You may need to implement a custom comparison function using https://pkg.go.dev/google.golang.org/protobuf/proto#Equal. 

This change needs tests that validates the behavior of the `DynamicFlowProcessor`

rectified95 replied:
Just checked under debugger that the update path is hit for a new config with only different filter contents (this means the custom comparators aren't needed in the dynamic exporter).

rectified95 replied:
The test `TestHandlersUpdatedInDfpOnConfigChange` checks the state of the `DynamicFlowProcessor` for a sequence of config reloads that register and unregister metrics and add filters with `assertHandlersInDfp`. I will also add the case I described above (filter modification). What else did you have in mind?

chancez replied:
Ah I missed that update. I was looking after the first but didn't see it due to how files get ordered in the diffs. Thanks. Will take a look t that.

chancez commented on pkg/hubble/metrics/dynamic_flow_processor.go:
```diff
@@ -0,0 +1,139 @@
+						d.logger.WithField("name", cm.Name).WithError(err).Error("HandleConfigurationUpdate failed for handler")
+					}
+					h.MetricConfig = cm
+				}
```
I'm not sure we should proceed if handling updates fails. Should we even proceed with the rest of the code? We probably don't want to update d.Metrics` if we get any errors, do we?

rectified95 replied:
Our validation functions rejects invalid config set reload attempts, eg. malformed params, or label changes - we take the all-or-nothing approach there. So when this code here executes and fails, that's currently due to unexpected Prometheus failures.

With errors only coming from Prometheus, we can't really have transactionality of config application, eg. for a this of 2 metrics [1: delete, 2: add]  if handler no.2 fails to register we can't cleanly undo no.1 having unregistered, since its metrics vectors will have been removed - we could only re-register it in an attempt to restore original state prior to applying the new config.

Since reverting to a clean state after partial failure cannot be achieved (can it?), I'd go for a best-effort approach of applying each config set that passed our validation.

chancez replied:
Right, well, I guess that's as good as it gets.

chancez commented on pkg/hubble/metrics/dynamic_flow_processor.go:
```diff
@@ -0,0 +1,139 @@
+		// This needs to happen first to properly check for conflicting plugins later during registration.
+		for _, m := range d.Metrics {
+			if _, ok := configuredMetricNames[m.Name]; !ok {
+				h := curHandlerMap[m.Name]
+				err := h.Handler.Deinit(d.registry)
+				if err != nil {
+					d.logger.WithField("name", m.Name).WithError(err).Error("Deinit failed for handler")
+				}
+				delete(curHandlerMap, m.Name)
```
Error handling is tricky. If a plugin fails to `Deinit` then it might be in an unclean state and then we might not want to proceed with updating the metrics config. However, at least with the currently written plugins, they only handle unregistering metrics so this is probably not an issue at the moment, but I worry about the repercussions if they ever have some kind of state.

Can you think of any good ways to handle potential error cases?

rectified95 replied:
I'd say the handlers are already stateful because they hold the config and metric vectors, so do you mean a situation where they hold state across re-registrations?

IMO a handler that failed to Deinit can be in an invalid state and we don't want to rely on it for reporting, so it should be removed from the map.

At the same time, if a failed Deinit leaves a handler in a state that still generates metrics, then perhaps it shouldn't be removed from the map here - we could decide based on the specific error returned. 

The question of transactionality of config changes for a failed Deinit seems the same as for a failed update below - I'll address it in your other comment.

tommyp1ckles commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -210,6 +210,7 @@ cilium-agent [flags]
       --hubble-drop-events                                        Emit packet drop Events related to pods (alpha)
```
One bit of feedback I have is how large this commit is, could you see if you could break this down into several smaller commits to make review, diffing and backporting easier?

See: [Cilium contribution docs section under how to submit a pull request.](https://docs.cilium.io/en/latest/contributing/development/contributing_guide/#submitting-a-pull-request)

rectified95 replied:
@tommyp1ckles I don't see an easy way to divide this into standalone and functional commits - I assume you refer to the first commit of the PR, not the feedback refactoring ones. Perhaps the below will be helpful in reviewing:

- The crux of the change is contained in 2 files: `dynamic_flow_processor.go` and `metric_config_watcher.go`, conceptually based on the `Dynamic Flowlog Exporter`. 
- The static/dynamic metrics fork happens in `hubbleintegration.go`
- All the handler files contain the same mechanical changes of setting new fields in the handler constructor, and implementng `Deinit`/`Update` functions.
- The watcher and the ConfigMap boilerplate follows directly the exporter code. 
- The Hubble flags pattern mirrors the other config toggles.

Let me know if this or the (updated) PR description is helpful, I'm happy to provide more details to make this easier for you. Perhaps a quick Slack chat will be the easiest.

tommyp1ckles replied:
@rectified95 

There's no hard and fast rule for commit size, but 1700 anything that's not all generated/vendored code changes does seem like a lot.  Specifically, I worry this might make identifying changes that potentially caused regressions more difficult.

Here's how I usually approach this (specifically, in the case of a PR that introduces new functionality):

* Flag & generated CLI documentation changes can be in their own commit.
* Put helm changes in their own commit.
* Primarily new code, such as that in `pkg/hubble/metrics/metric_config_watcher.go` and `pkg/hubble/metrics/dynamic_flow_processor.go` and related test changes can probably be in a seperate commit while not breaking the per-commit build rule.
* Finally, in-line changes integrating the new code into existing code can be their own commits.  As a reviewer, I find these ones particularly helpful as the commit message and smaller set of changes can make it easier to understand the broader impact of the change.

rectified95 replied:
@tommyp1ckles Per your suggestion, I've split this PR into several commits: docs, Helm, new code, integration code.
To ensure each one builds, I added empty implementations of the new interface methods in the intermediate commits.
Since this operation undoes earlier squashing, it was at times tricky - I ended up putting the test code with the integration commit since it relies on its final shape post-refactoring. 

Changelist for the integration commit (here, since I've seen the CI reject commit messages over 130 characters.. unless that was just the title line):
- fork the metrics setup path in Hubble launch to use the dynamic config path
- remove the `FlowProcessor` interface and move its `ProcessFlow` method into the `Handler` interface (rationale in PR description + this thread: https://github.com/cilium/cilium/pull/35185#discussion_r1797444382)
- remove the wrapper type `Handlers` and its pass-through `ProcessFlow` function, now that we don't maintain the list of `FlowProcessor`s
- convert handler bulk creation functions to operate on a single handler and be iterated on (needed for DFP applying metric configs one by one)
- adapt existing tests to use refactored types

The resulting diff from the force-push contains 3 changes:
- empty line removal 
- removed commented import line
- removed redundncy I spotted in DFP: it already creates the map it was calling a helper for

https://github.com/cilium/cilium/compare/0658165559d45752ebaa6310d990bb43be1ebe8b..af5ba6603351a936345f498bd46ca144930378bc

chancez reviewed (approved):
Just did a re-review, especially focusing on the new tests and everything looks good to me. Thanks!

squeed reviewed (approved):
Helm bits look good, thanks!

tommyp1ckles commented on pkg/hubble/metrics/dynamic_flow_processor.go:
```diff
@@ -0,0 +1,134 @@
+	if err != nil {
+		panic(err)
+	}
+	// TODO don't panic, add transaction recovery logic
```
Leftover TODOS/comments

rectified95 replied:
This one is not in the diff for the current commit set - must have been an older commit before the approvals and rebases.

rectified95 committed:
add flags, generated docs

Adds new Hubble flag "hubble-dynamic-metrics-config-path".

Signed-off-by: Igor Klemenski <igor.klemenski@microsoft.com>

rectified95 committed:
add helm schema and ConfigMap

Sets up new ConfigMap "cilium-dynamic-metrics-config" and new Helm options under ".Values.hubble.metrics.dynamic".

Signed-off-by: Igor Klemenski <igor.klemenski@microsoft.com>

rectified95 committed:
add DynamicFlowProcessor

Adds new logic in Hubble for dynamic metric registration.
Adds new methods to the Handler interface: HandleConfigurationUpdate, Deinit.

Signed-off-by: Igor Klemenski <igor.klemenski@microsoft.com>

rectified95 committed:
add filters to handlers

Pass filters to Hubble metric handlers' Init method.

Signed-off-by: Igor Klemenski <igor.klemenski@microsoft.com>

rectified95 committed:
integrate DFP, remove FlowProcessor, add tests

Fork metric setup logic into static and dynamic paths during Hubble launch.
Split handler bulk creation methods into single iterators.
Move ProcessFlow method into the Handler interface.

Signed-off-by: Igor Klemenski <igor.klemenski@microsoft.com>

tommyp1ckles commented on pkg/hubble/metrics/metric_config_watcher.go:
```diff
@@ -0,0 +1,154 @@
+	if c.ticker != nil {
+		c.ticker.Stop()
+	}
+	c.stop <- true
```
It probably makes sense to just close the channel here.

rectified95 replied:
~~If I do that, I'll need to check for the closure instead of consuming this value - is there much difference?
Taking a quick look here (https://gobyexample.com/closing-channels) this seems idiomatic (?)~~
EDIT: closing the channel now

rectified95 replied:
I am just realizing that the DynamicFlowProcessor is missing its `Stop` method that would call this one in response to a context closure event and stop the watcher.. Let me fix that up.
@chancez FYI Should we also `Deinit` all handlers there?

chancez replied:
We could add a Shutdown method to the `DynamicFlowExporter` to stop the watcher for sure. As far as running `Deinit` for all handlers in shutdown, I think that feels like a reasonable behavior.

rectified95 replied:
Cool, added in this commit: https://github.com/cilium/cilium/pull/35185/commits/2bda66ee05b35edbfa3920ac44d6416b20631568

tommyp1ckles commented on pkg/hubble/metrics/metric_config_watcher.go:
```diff
@@ -0,0 +1,154 @@
+	defer c.mutex.Unlock()
+
+	config := &api.Config{Metrics: []*api.MetricConfig{}}
+	yamlFile, err := os.ReadFile(c.configFilePath)
```
One thing I'm wondering about, in most cases where Cilium is checking for changes to a file we use the `pkg/fswatcher` package instead of polling. Could that make sense in this instance?

rectified95 replied:
I think it could, perhaps as a future improvement. For now, I just followed the example of the dynamic exporter.
Compared to the exporter, this code also bails earlier when it detects the file hash has not changed, rather than unmarshalling the yaml first.

Moreover, I noticed problems when trying this API: text editors will issue multiple update events before a file is saved. So unless we can guarantee the configMap update is always done atomically by the user, I'd rather keep it as is for now. How does that sound?

https://github.com/fsnotify/fsnotify?tab=readme-ov-file#watching-a-file-doesnt-work-well

rolinh commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -993,6 +993,10 @@ data:
 {{- end }}
 {{- end }}
+{{- if .Values.hubble.metrics.dynamic.enabled }}
+  hubble-dynamic-metrics-config-path: /dyn-metrics-config/dyn-metrics.yaml
```
Is there a good reason for using both "dynamic" and the shortened version "dyn"? If not, I'd suggest to stick to "dynamic" everywhere.

rectified95 replied:
Done

rolinh commented on pkg/hubble/cell/hubbleintegration.go:
```diff
@@ -335,16 +335,42 @@ func (h *hubbleIntegration) launch(ctx context.Context) {
+				h.log.WithError(err).Error("Unable to setup metrics: %w", err)
+				return
+			}
 
```
What about using switch case to make this whole section a little bit more readable?

rectified95 replied:
Done

rolinh commented on pkg/hubble/metrics/dynamic_flow_processor.go:
```diff
@@ -0,0 +1,151 @@
+		if m, ok := curHandlerMap[cm.Name]; ok {
+			if reflect.DeepEqual(*m.MetricConfig, *cm) {
+				continue
+			} else {
```
This else clause is superfluous, removing it would make the code more readable. In general, we should try to keep the code path at minimal indentation.

rectified95 replied:
Removed 2 levels of indentation - the `if` inside this `else`was duplicating the outer `if.`

rectified95 committed:
stop watcher on ctx.Done(), reduce indentation

Stop Hubble dynamic metric config change watcher when context is closed.
Return early in Hubble start when both dynamic and static metrics specified.
Fix pointer ownership convention.

Signed-off-by: Igor Klemenski <igor.klemenski@microsoft.com>

tommyp1ckles reviewed (approved):
CLI lgtm, thanks for bearing with me on the commits 😃  @rectified95

rolinh commented on pkg/hubble/cell/hubbleintegration.go:
```diff
@@ -341,16 +341,49 @@ func (h *hubbleIntegration) launch(ctx context.Context) {
+		switch {
+		case h.config.DynamicMetricConfigFilePath != "" && len(h.config.Metrics) > 0:
+			{
+				h.log.Error("Cannot configure both static and dynamic Hubble metrics")
```
Should we return here? Not being able to expose metrics should be reason enough to not start Hubble (or fail the metrics sub-system once it's a cell). At the very least, let's add a comment so that we can address that once we refactor Hubble metrics into its own cell (#34501).

rectified95 replied:
Returning here makes sense to me, that's what the code does in the static case currently anyway.

rolinh commented on pkg/hubble/cell/hubbleintegration.go:
```diff
@@ -341,16 +341,49 @@ func (h *hubbleIntegration) launch(ctx context.Context) {
-			return
+				observerOpts = append(observerOpts,
+					observeroption.WithOnDecodedFlowFunc(func(ctx context.Context, flow *flowpb.Flow) (bool, error) {
+						if metrics.EnabledMetrics != nil {
```
We're in the case for static metrics here, right? It sounds like this clause should be part of the switch case condition, not within the `OnDecodedFlowFunc`, ie:
```go
case metrics.EnabledMetrics != nil && len(h.config.Metrics) > 0:
```

rectified95 replied:
This is the static path, but this if-clause can’t be moved to the switch condition because `metrics.EnabledMetrics` is only initialized by `metrics.InitMetrics()` within this case block.
You could argue that the callback registered by `WithOnDecodedFlowFunc` will never encounter a nil value, but I left the check there unchanged for unexpected bad states.

rolinh commented on pkg/hubble/metrics/api/api.go:
```diff
@@ -64,56 +59,59 @@ type Handler interface {
 }
 
-// ProcessFlow processes a flow by calling ProcessFlow it on to all enabled
+func InitHandler(log logrus.FieldLogger, registry *prometheus.Registry, item *NamedHandler, handlers *[]NamedHandler) error {
```
Passing a pointer to a slice of `NamedHandler` seems like an anti-pattern to me here. We're probably better off returning the slice of `NamedHandler` alongside the error. Or better yet, move the body of this function into the `for` loop of `InitHandlers`.

rectified95 replied:
Changed this to append the initialized handler in the caller.
As for moving the body to `InitHandlers`: this function is needed for one-by-one metric creation in the `DynamicFlowProcessor`, so it has to stay.

rolinh commented on pkg/hubble/metrics/api/api.go:
```diff
@@ -64,56 +59,59 @@ type Handler interface {
+// ExecuteAllProcessFlow processes a flow by calling ProcessFlow on all enabled
 // metric handlers
-func (h Handlers) ProcessFlow(ctx context.Context, flow *pb.Flow) error {
+func ExecuteAllProcessFlow(ctx context.Context, flow *pb.Flow, handlers *[]NamedHandler) error {
```
I see no reason to pass a pointer to a slice here. Maybe more importantly, I don't think we _need_ to have this function in `metrics/api`. Keeping this simple for loop in both the dynamic and static metrics code paths would make the code much easier to understand. Sometimes, a little duplication is better than more abstractions.

rectified95 replied:
Ack: changed to use the for-loop in both static and dynamic code paths.
Moved this function out of the api for use as a test helper.

rolinh commented on pkg/hubble/metrics/api/api.go:
```diff
@@ -64,56 +59,59 @@ type Handler interface {
-	for _, h := range h.handlers {
-		for _, mv := range h.ListMetricVec() {
-			if ctx := h.Context(); ctx != nil {
+func ProcessCiliumEndpointDeletion(endpoint *types.CiliumEndpoint, handlers *[]NamedHandler) {
```
Ditto.

rectified95 replied:
Removed pointer from the signature.
Though used a lot of `registry_test.go`, this function looks _unused_ in product code.
Could remove, but I'd like to contain the scope of this PR. 
(There's an identically named function in `metrics.go` that does something different.

rolinh replied:
sgtm, this PR is indeed quite large already.

rolinh reviewed (changes requested):
Thanks for implementing the feedback @rectified95. I gave the patch another read and have some more feedback. Generally speaking, I find it difficult to understand what the code does. I left some suggestions for improvements, see below.

rolinh reviewed (approved):
Thanks @rectified95, lgtm. Would you mind creating an issue to track  some of the cleanup that this PR now allows around the API package? As discussed, this PR is indeed already pretty large in its own.

rolinh reviewed (changes requested):
Sorry, could you please squash your latest commit ("pr feedback") before we merge?

rolinh reviewed (approved):
LGTM, thanks!
